covid_graphs/covid_graphs/formula.py

```diff
@@ -1,81 +1,90 @@
+            plot_start_date + datetime.timedelta(days=d)
+            for d in range((plot_end_date - plot_start_date).days)
+        ]
+        self._ys = [self._func(x) for x in self._xs]
+        return self._xs, self._ys
 
+    def get_maximum(self) -> Tuple[datetime.date, float]:
+        """
+        Returns (x, y) point on the curve with maximal y.
+        """
+        if self._xs is None or self._ys is None:
+            raise ValueError("Must call get_trace() first.")
+        idx = np.argmax(self._ys)
+        return self._xs[idx], self._ys[idx]
 
-def first_day_of_curve(cumulative_active, formula):
-    """
-    Returns the first day of the curve formed by 'formula'.
 
-    It's the first day for which the number of cumulative active is more than
-    formula.min_case_count.
-    """
-    return np.argmax(cumulative_active > formula.min_case_count)
+class CurveConstructor:
+    @abstractmethod
+    def get_curve(self, country_report: CountryReport) -> Curve:
+        pass
 
 
-class Curve:
-    """
-    Class containing a curve derived from a formula.
+@dataclass
+class AtgCurveConstructor(CurveConstructor):
+    tg: float
+    a: float
+    exponent: float
+    min_case_count: int
 
-    cumulative_active - list of cumulative active cases
-    first_idx - 0-based index where the graph starts, so from cumulative_active[first_idx:].
-    last_idx - 0-based index where the graph ends
-    first_date - date corresponding to cumulative_active[0]
-    """
+    def get_curve(self, country_report: CountryReport) -> Curve:
+        label = r"$\frac{_A}{_TG} \cdot \left(\frac{t}{_TG}\right)^{_expon} / e^{t/_TG}$"
+        label = (
+            label.replace("_A", f"{self.a:.0f}")
+            .replace("_TG", f"{self.tg}")
+            .replace("_expon", f"{self.exponent}")
+        )
 
-    def __init__(self, formula, cumulative_active, first_idx, last_idx, first_date):
-        self.text = formula.text
-        start_idx = first_day_of_curve(cumulative_active, formula)
-        length = last_idx - start_idx + 1
-        self.y = formula.lambd(np.arange(length) + 1)
-        self.t = np.arange(length) + 1 + (start_idx - first_idx)
-        self.date_list = [
-            (first_date + timedelta(days=d)).strftime("%Y-%m-%d")
-            for d in range(start_idx, last_idx + 1)
-        ]
+        # Display values from the first day for which the number of cumulative active is
+        # at least min_case_count. This day is also "day one".
+        start_idx = np.argmax(country_report.cumulative_active >= self.min_case_count)
+        start_date = country_report.dates[start_idx]
+        day_zero = start_date - datetime.timedelta(days=1)
 
-        self.maximal_y = self.y.max()
-        self.maximal_idx = self.y.argmax()
+        # Display values until the second inflection point.
+        second_ip_days = math.ceil(self.tg * (self.exponent + math.sqrt(self.exponent)))
+        end_date = start_date + datetime.timedelta(days=second_ip_days)
```

Comment:
`end_date` is perhaps not the best name, since the curve doesn't really have an end and we might override it anyway. A possible name is `second_ip_day`, but probably not the best.

We clearly want to show it at least up to the second inflection point. I'm tired and out of good ideas, but something like `show_at_least_to_date` would be better. I can't think of a shorter name right now.